Improve remote-externalities

hardliner66 opened the pull request:
Fixes parts of #8379.

This PR changes the CLI parameters for try-runtime. It adds the capability to write to a cache file when using try-runtime with the `live` state. Additionally it adds the capabillity to specify the modules to scrape and the block number from which to start.

It also changes `remote_externalities::Builder` to be generic over the `Block` type.

I still need to address the broken tests that are mentioned in #8379, but everything else should already work.

niklasad1 commented on utils/frame/try-runtime/cli/src/lib.rs:
```diff
@@ -60,31 +56,74 @@ pub struct TryRuntimeCmd {
-	/// A live chain. Inner value is the HTTP uri.
-	Live(String),
+fn parse_url(s: &str) -> Result<String, &'static str> {
+	match s.get(..7) {
```
I think `s.starts_with("http://")`is more readable in this case

hardliner66 replied:
I changed the code to use starts_with.

niklasad1 commented on utils/frame/try-runtime/cli/src/lib.rs:
```diff
@@ -123,11 +166,27 @@ impl TryRuntimeCmd {
+						name: c.file_name.clone(),
+						directory: c.directory.clone(),
+					}),
+					modules: modules.clone().unwrap_or(Vec::new()),
```
```suggestion
					modules: modules.clone().unwrap_or_default()
```

kianenigma commented on utils/frame/remote-externalities/src/lib.rs:
```diff
@@ -115,27 +115,35 @@ use sp_core::{
 use jsonrpsee_http_client::{HttpClient, HttpConfig};
 
+use sp_runtime::{
+	generic::BlockId,
+    traits::{
+		Block as BlockT, NumberFor,
+    }
```
```suggestion
	generic::BlockId,
    traits::{Block as BlockT, NumberFor}
```

kianenigma commented on utils/frame/remote-externalities/src/lib.rs:
```diff
@@ -234,10 +242,10 @@ impl Builder {
+		RpcApi::<B>::finalized_head(&self.as_online().rpc()).await.map_err(|e| {
 			error!("Error = {:?}", e);
 			"rpc finalized_head failed."
 			})
```
```suggestion
		})
```

kianenigma commented on utils/frame/remote-externalities/src/lib.rs:
```diff
@@ -249,18 +257,30 @@ impl Builder {
+		RpcApi::<B>::storage_pairs(&self.as_online().rpc(), prefix, Some(at)).await.map_err(|e| {
 			error!("Error = {:?}", e);
 			"rpc storage_pairs failed"
 			})
```
```suggestion
		})
```

kianenigma commented on utils/frame/remote-externalities/src/lib.rs:
```diff
@@ -249,18 +257,30 @@ impl Builder {
+		RpcApi::<B>::block_hash(&self.as_online().rpc(), number).await.map_err(|e| {
+			error!("Error = {:?}", e);
+			"rpc block_hash failed"
+			})
```
```suggestion
		})
```

kianenigma commented on utils/frame/remote-externalities/src/lib.rs:
```diff
@@ -379,6 +406,9 @@ impl Builder {
 #[cfg(test)]
 mod tests {
 	use super::*;
+	pub type Header = sp_runtime::generic::Header<u32, sp_runtime::traits::BlakeTwo256>;
```
You could also just use these: https://crates.parity.io/sp_runtime/testing/index.html

hardliner66 replied:
I replaced them with the types in sp_runtime::testing

kianenigma commented on utils/frame/try-runtime/cli/src/lib.rs:
```diff
@@ -60,31 +56,74 @@ pub struct TryRuntimeCmd {
+		#[structopt(short, long)]
+		cache_file: Option<CacheParams>,
+
+		/// The block number at which to connect. Can be either a number or a hash (staring with `0x`) Will be latest finalized head if not provided.
```
you might need to break these lines if they are more than 100 cols

https://github.com/paritytech/substrate/blob/master/docs/STYLE_GUIDE.md

kianenigma commented on utils/frame/try-runtime/cli/src/lib.rs:
```diff
@@ -60,31 +56,74 @@ pub struct TryRuntimeCmd {
+	match s.get(..7) {
+		// could use Url crate as well, but lets keep it simple for now.
+		Some("http://") => Ok(s.to_string()),
+		_ => Err("not a valid url"),
```
```suggestion
		_ => Err("not a valid HTTP url: must start with 'http://'"),
```

kianenigma reviewed (commented):
I'll give another detailed review next week, loosk great so far.

hardliner66 commented:
I removed the ability to use a number for the block number as it was rather difficult to get the corresponding hash, so now only hashes are allowed.

I also replaced async-std with tokio, to make the tests runnable. The live tests now all succeed except the one with the cache file. Here I still need to figure out what the best way to produce a small cache file is and add it to the repo.

hardliner66 commented:
I added a small cache file and moved the test that uses it out of the feature gated test module. Now all tests are succeeding.

dvdplm commented on utils/frame/remote-externalities/src/lib.rs:
```diff
@@ -375,49 +376,62 @@ impl Builder {
-mod tests {
-	use super::*;
+mod test_prelude {
+	pub use super::*;
```
Best limit the visibility of these items.

```suggestion
	pub(crate) use super::*;
```
…and the same for the others.

dvdplm commented on utils/frame/remote-externalities/src/lib.rs:
```diff
@@ -375,49 +376,62 @@ impl Builder {
 
-	#[async_std::test]
-	async fn can_load_cache() {
+#[cfg(feature = "remote-test")]
+#[cfg(test)]
```
Maybe:
```suggestion
#[cfg(all(test, feature = "remote-test"))]
```

…although it's a tossup which is more readable. :)

dvdplm commented on utils/frame/remote-externalities/src/lib.rs:
```diff
@@ -375,49 +376,62 @@ impl Builder {
 			}))
 			.build()
 			.await
 			.unwrap()
```
Not your code but I think an `.expect("Can't reach the remote node. Is it running?")` or something to that effect might be useful to forgetful devs, here and for the other two tests? Wdyt?

hardliner66 replied:
Good idea. I also added an error message to the offline test.

dvdplm commented on utils/frame/try-runtime/cli/src/lib.rs:
```diff
@@ -60,31 +56,93 @@ pub struct TryRuntimeCmd {
+
+	/// Use a live chain to run the migration.
+	Live {
+		/// An optional cache file to WRITE to. Not cached if set to `None`.
```
@kianenigma I'm not completely sold on the naming here. A cache to me is something that optimize access to some resource and is not necessary for the logic to work; the real resource remains available when the cache is unable to provide the data. This is more of a data file containing state data. 

`state_file`?

kianenigma replied:
> A cache to me is something that optimize access to some resource and is not necessary for the logic to work; the real resource remains available when the cache is unable to provide the data. 

The current system also matches your description. In `Live` mode, we always use the remote node, so the resource is there. This is about updating/creating a cache file. 

I am okay with renaming `cache` everywhere to `state_snapshot`.

hardliner66 replied:
I renamed cache to state_snapshot in remote-externalities. In try-runtime I changed it to snapshot_path because that seems more fitting for the cli. If you like I can change it to state_snapshot as well.

dvdplm commented on utils/frame/try-runtime/cli/src/lib.rs:
```diff
@@ -60,31 +56,93 @@ pub struct TryRuntimeCmd {
+
+	if let Some(pos) = block_number.chars().position(|c| !c.is_ascii_hexdigit()) {
+		Err(format!(
+			"Expected block hash, found illegal hex character at position: {}",
```
Nice, I like the effort to provide users with a good error message.

dvdplm commented on utils/frame/try-runtime/cli/src/lib.rs:
```diff
@@ -60,31 +56,93 @@ pub struct TryRuntimeCmd {
+}
+
+fn parse_url(s: &str) -> Result<String, &'static str> {
+	if s.starts_with("http://") {
```
What about `https://`? I think that should be supported too.

kianenigma replied:
I don't know if our nodes even expose https endpoints, do they? 

Either way, this is rooted in my old code. Back in the days, the `ws/wss` endpoints where not capable of delivering the whole state, and only HTTP could.

dvdplm replied:
> I don't know if our nodes even expose https endpoints, do they?

I bet most production deployments stick a TLS proxy in front of their nodes. And we will support https directly in `jsonrpsee`. :)

dvdplm commented on utils/frame/try-runtime/cli/src/lib.rs:
```diff
@@ -60,31 +56,93 @@ pub struct TryRuntimeCmd {
+
+fn parse_url(s: &str) -> Result<String, &'static str> {
+	if s.starts_with("http://") {
+		// could use Url crate as well, but lets keep it simple for now.
```
FWIW `url` is already a dependency elsewhere in `substrate`. If you don't want to do it in this PR I suggest we remove this comment, doesn't add much imo. :)

dvdplm commented on utils/frame/try-runtime/cli/src/lib.rs:
```diff
@@ -60,31 +56,93 @@ pub struct TryRuntimeCmd {
 
-impl FromStr for State {
+#[derive(Debug, structopt::StructOpt)]
+pub struct CacheParams {
```
What do you think of calling this `SnapshotPath`?

Further down we have:
```
			State::Snap {
					cache_params: CacheParams { file_name, directory }
				} => … … …
```

Wouldn't this read a bit better: `State::Snap { path: SnapshotPath { file_name, directory }`?

kianenigma replied:
Related to https://github.com/paritytech/substrate/pull/8397/files#r598078128, mostly relates to my bad original naming, and agree to change.

dvdplm commented on utils/frame/try-runtime/cli/src/lib.rs:
```diff
@@ -60,31 +56,93 @@ pub struct TryRuntimeCmd {
+		match file_name {
+			Some(file_name) => Ok(Self {
+				// TODO: maybe don't use to_string_lossy here
+				directory: parent.map(|p| p.to_string_lossy().into()).unwrap_or(".".to_string()),
```
I think it's a bit dangerous to let an invalid path through here. A bit contrived perhaps, but if `file_name` is present both in `.` and whatever path I thought I gave then something bad could happen?

kianenigma replied:
actually there's a TODO in `CacheConfig` as 

```
	// TODO: I could mix these two into one filed, but I think separate is better bc one can be
	// configurable while one not.
```

I also thought about the dangers. You are potentially giving write primitive to an external cli input. Technically, a malicious input can overwrite some critical file for example. But the thing is, no one is going to expose this functionality in the wild or anything. It will all be within the trust boundary, so I think it is fine.

kianenigma replied:
That being said, sanitizing the path is a good idea anyway.

hardliner66 replied:
I think it might be a good idea to use `PathBuf` internally. This way we can at least be sure that no chars will be dropped during the string conversion. And we can expose the `PathBuf` directly as a structopt parameter, which would remove the need for manual conversions at all.

I also think it might would be better to combine directory and filename into one field. Internally it is combined to a single `PathBuf` anyway. If we still want the directory and path to be set separately, we can still do this on the cli level, but then it also needs to be validated that the file_name is always just a file and doesn't contain directory parts. Otherwise the directory parameter would be redundant.

hardliner66 replied:
@kianenigma Do you want me to combine the paths into a single `PathBuf` for this PR or should that be handled separately?

kianenigma replied:
I prefer not having major changes anymore once the PR is approves, but given the simplicity of this, whichever your find better, both fine by me.

hardliner66 replied:
Okay, I'm gonna prepare a separate PR then.

dvdplm commented on utils/frame/try-runtime/cli/src/lib.rs:
```diff
@@ -93,6 +151,10 @@ impl TryRuntimeCmd {
 	where
 		B: BlockT,
+		B::Hash: FromStr,
+		<B::Hash as FromStr>::Err: std::fmt::Debug,
```
`Debug` is already in scope.
```suggestion
		<B::Hash as FromStr>::Err: Debug,
```

dvdplm commented on utils/frame/try-runtime/cli/src/lib.rs:
```diff
@@ -123,11 +185,31 @@ impl TryRuntimeCmd {
+					cache: CacheConfig {
+						name: file_name.into(),
+						directory: directory.into(),
+						..Default::default()
```
Isn't `..Default::default` superfluous?

kianenigma replied:
`CacheConfig` has only two fields, yeah

dvdplm commented on utils/frame/try-runtime/cli/src/lib.rs:
```diff
@@ -123,11 +185,31 @@ impl TryRuntimeCmd {
+					modules
+				} => Builder::<B>::new().mode(Mode::Online(OnlineConfig {
+					uri: url.into(),
+					cache: cache_file.as_ref().map(|c| CacheConfig {
```
Having a `CacheConfig::new(file, dir)` might be worth considering.

kianenigma commented on utils/frame/remote-externalities/src/lib.rs:
```diff
@@ -115,27 +115,28 @@ use sp_core::{
 
+use sp_runtime::traits::Block as BlockT;
+
+// TODO: Make KeyPair generic
```
hmm? really? :D

hardliner66 replied:
The original todo was "make these two generic." and I assumed it was meant for both the hash and the keypair. But I only made the hash generic, so I re-added the todo for the keypair, so it wont get lost.

kianenigma replied:
ahh okay, we can remove this. Storage key and pair are already Vec<u8>, generic enough.

kianenigma replied:
```suggestion
```

kianenigma commented on utils/frame/remote-externalities/src/lib.rs:
```diff
@@ -375,49 +376,62 @@ impl Builder {
 			.build()
 			.await
 			.unwrap()
 			.execute_with(|| {});
```
My original tests were not really sophisticated, and I think you can do better here: 

We need a way to assert that _some_ data has _actually_ been injected into to the proxy pallet. For example, if you scrape the system pallet, now you want to try and read something like `System::block_number` or some other storage item that is unlikely not to change. 

Or, you can assume that you know all storage items in proxy are prefixed with `twox128("Proxy")`, and then assert that `sp_io::storage::next_key(twox128("Proxy"))` is some. 

There are other ways to also do this. 

Treat it as a bonus, if you want to explore the storage a bit.

kianenigma commented on utils/frame/try-runtime/cli/src/lib.rs:
```diff
@@ -60,31 +56,93 @@ pub struct TryRuntimeCmd {
+		#[structopt(short, long)]
+		cache_file: Option<CacheParams>,
+
+		/// The block hash at which to connect.
```
you support number as well, doc is outdated.

hardliner66 replied:
I removed number support, because I had some problems with it. So we're back to only hashes.

kianenigma replied:
ahh okay. I just realized that as well.

kianenigma commented on utils/frame/try-runtime/cli/src/lib.rs:
```diff
@@ -60,31 +56,93 @@ pub struct TryRuntimeCmd {
+		let parent = p.parent();
+		let file_name = p.file_name();
+
+		match file_name {
+			Some(file_name) => Ok(Self {
+				// TODO: maybe don't use to_string_lossy here
+				directory: parent.map(|p| p.to_string_lossy().into()).unwrap_or(".".to_string()),
+				file_name: file_name.to_string_lossy().into()
+			}),
+			None => Err("invalid path"),
```
```suggestion
		file_name.and_then(|file_name| Self {
			directory: parent.map(|p| p.to_string_lossy().into()).unwrap_or(".".to_string()),
			file_name: file_name.to_string_lossy().into()
		}).ok_or("invalid path")
```

kianenigma commented on utils/frame/try-runtime/cli/src/lib.rs:
```diff
@@ -123,11 +185,31 @@ impl TryRuntimeCmd {
 			let builder = match &self.state {
-				State::Snap(file_path) => Builder::new().mode(Mode::Offline(OfflineConfig {
-					cache: CacheConfig { name: file_path.into(), ..Default::default() },
+				State::Snap {
```
Can this be reformatted? looks unlike a normal match to me.

kianenigma reviewed (approved):
Some renaming nits remaining, but looks good to me otherwise, thanks!

dvdplm reviewed (approved):
lgtm.

dvdplm commented:
@kianenigma I'm hazy on the labelling rules here, `B0` for this one yes?

kianenigma commented:
done @dvdplm

dvdplm commented:
Let's avoid merging this until @hardliner66 has had time to make a few changes more. @hardliner66 ping us for a last once-over whenever you're ready, ok? :)

hardliner66 commented:
@dvdplm After talking with @kianenigma I decided to create a separate pull request with the remaining changes. So the PR can be merged.
